packages/enzyme/src/selectors.js

```diff
@@ -0,0 +1,318 @@
+import unique from 'lodash/uniq';
+import {
+  treeFilter,
+  nodeHasId,
+  findParentNode,
+  nodeMatchesObjectProps,
+  childrenOfNode,
+  hasClassName,
+} from './RSTTraversal';
+import { nodeHasType, nodeHasProperty } from './Utils';
+// our CSS selector parser instance
+const parser = createParser();
+
+// Combinators that allow you to chance selectors
+const CHILD = 'childCombinator';
+const ADJACENT_SIBLING = 'adjacentSiblingCombinator';
+const GENERAL_SIBLING = 'generalSiblingCombinator';
+const DESCENDANT = 'descendantCombinator';
+
+// Selectors for targeting elements
+const SELECTOR = 'selector';
+const TYPE_SELECTOR = 'typeSelector';
+const CLASS_SELECTOR = 'classSelector';
+const ID_SELECTOR = 'idSelector';
+const ATTRIBUTE_PRESENCE = 'attributePresenceSelector';
+const ATTRIBUTE_VALUE = 'attributeValueSelector';
+// @TODO we dont support these, throw if they are used
+const PSEUDO_CLASS = 'pseudoClassSelector';
+const PSEUDO_ELEMENT = 'pseudoElementSelector';
+
+/**
+ * Calls reduce on a array of nodes with the passed 
+ * function, returning only unique results.
+ * @param {Function} fn 
+ * @param {Array<Node>} nodes 
+ */
+function uniqueReduce(fn, nodes) {
+  return unique(nodes.reduce(fn, []));
+}
+
+/**
+ * Takes a CSS selector and returns a set of tokens parsed
+ * by scalpel.
+ * @param {String} selector 
+ */
+function safelyGenerateTokens(selector) {
+  try {
+    return parser.parse(selector);
+  } catch (err) {
+    throw new Error(`Failed to parse selector: ${selector}`);
+  }
+}
+
+/**
+ * Takes a node and a token and determines if the node
+ * matches the predicate defined by the token.
+ * @param {Node} node 
+ * @param {Token} token 
+ */
+function nodeMatchesToken(node, token) {
+  if (node === null || typeof node === 'string') {
+    return false;
+  }
+  switch (token.type) {
+    /**
+     * Match against the className prop
+     * @example '.active' matches <div className='active' />
+     */
+    case CLASS_SELECTOR:
+      return hasClassName(node, token.name);
+    /**
+     * Simple type matching
+     * @example 'div' matches <div />
+     */
+    case TYPE_SELECTOR:
+      return nodeHasType(node, token.name);
+    /**
+     * Match against the `id` prop
+     * @example '#nav' matches <ul id="nav" />
+     */
+    case ID_SELECTOR:
+      return nodeHasId(node, token.name);
+    /**
+     * Matches if an attribute is present, regardless
+     * of its value
+     * @example '[disabled]' matches <a disabled />
+     */
+    case ATTRIBUTE_PRESENCE:
+      return nodeHasProperty(node, token.name);
+    /**
+     * Matches if an attribute is present with the
+     * provided value
+     * @example '[data-foo=foo]' matches <div data-foo="foo" />
+     */
+    case ATTRIBUTE_VALUE:
+      return nodeHasProperty(node, token.name, token.value);
+    case PSEUDO_ELEMENT:
+    case PSEUDO_CLASS:
+      throw new Error('Enzyme::Selector does not support psuedo-element or psuedo-class selectors.');
```

Comment:
Adding support for pseudo-selectors will be really easy. For example, implementing `:not` (https://github.com/airbnb/enzyme/pull/460). I figure this can be done in a follow-up.

Reply:
Followups for these things are great; this PR's big enough already :-)